src/typechecker.erl

```diff
@@ -2302,7 +2302,17 @@ do_type_check_expr_in(Env, ResTy, {call, P, Name, Args}) ->
 do_type_check_expr_in(Env, ResTy, {lc, P, Expr, Qualifiers}) ->
     type_check_comprehension_in(Env, ResTy, lc, Expr, P, Qualifiers);
 do_type_check_expr_in(Env, ResTy, {bc, P, Expr, Qualifiers}) ->
-    type_check_comprehension_in(Env, ResTy, bc, Expr, P, Qualifiers);
+    Error = {type_error, bc, P, ResTy},
+    {Res, Acc} = lists:mapfoldl(
+        fun(FlatTy, Acc) ->
+            case catch {type_check_comprehension_in(Env, FlatTy, bc, Expr, P, Qualifiers), Acc} of
+                {{#{}, C1}, {#{}, C2}}                 -> {ok, {#{}, constraints:combine(C1, C2)}};
+                {{#{}, _} = C, {type_error, _, _, _}} -> {ok, C};
+                {type_error, _, _, FlatTy}            -> {type_error,Acc}
+            end
+        end, Error, flatten_type(ResTy, Env)),
```

Comment:
I'm catching exactly on `{type_error...}` because that is the only thing that is thrown on the callees while exploring the code. Also, if there's any other error, this is not where I want to catch it. But I've realised I might be masking different `{type_error,...` messages being thrown and making them all be only the main one... :thinking: 

About `flatten_type/2`, it is used in normalised, but then normalised also does a `case merge_union_types/2 of` which, when matching on a list longer than one, returns again a type-union of them (having removed possible duplicates and so on, but in this case it means nothing changed. That's why I directly used `flatten_type/2`, I want the list to iterate through.

Also, where do you mean we can move the logic? To `do_type_check_expr_in` caller, that is, the ubiquitous `type_check_expr_in`, right? Hmm... this is getting a lot more complex than I expected as an entry point for me, but well, sounds cool :smile:

Reply:
> I'm catching exactly on {type_error...} because that is the only thing that is thrown on the callees while exploring the code. Also, if there's any other error, this is not where I want to catch it.

That's not true. You're catching all kinds of exceptions and then crashing on a case clause.

This is what's happening in a catch without try:

```Erlang
1> Thrower = fun() -> case catch throw(boomerang) of [] -> [] end end.
#Fun<erl_eval.20.128620087>
2> try Thrower() catch C:T:E -> io:format("~p:~p~n at ~p~n", [C, T, E]) end.
error:{case_clause,boomerang}
 at [{erl_eval,case_clauses,6,[{file,"erl_eval.erl"},{line,968}]},
     {erl_eval,try_clauses,8,[{file,"erl_eval.erl"},{line,914}]},
     {shell,exprs,7,[{file,"shell.erl"},{line,686}]},
     {shell,eval_exprs,7,[{file,"shell.erl"},{line,642}]},
     {shell,eval_loop,3,[{file,"shell.erl"},{line,627}]}]
ok
```

To catch selectively, you must use try-catch, not only catch.

> That's why I directly used flatten_type/2, I want the list to iterate through.

I understand it does exactly what you need here, but the name is not very comprehensive and there is no documentation for it. Please rename it! If it has a comprehensive name, it can be used and I can remove it from my list of functions to move and make unexported in another modules. Please rename. (I'm just trying to line out some sort of modularization of the giant module.)

> Also, where do you mean we can move the logic?

`type_check_expr_in` is normalizing the type and then calls its helper `do_type_check_expr_in`. Put the function clause `do_type_check_expr_in(Env, {type, _, union, Tys}, Expr)` before any function clauses dissecting Expr. Then it can safely, recursively call itself, because there is no more union on the same level.

One more thing: Please, don't do as I suggest if you don't agree! It's just comments. :-)